### PR TITLE
feat: add `ignoreResourceUpdates` to reduce controller CPU usage (#13534)

### LINKED ISSUE #13534: Unnecessary application refresh on update event cause high CPU usage on controller
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

TLDR; When a kubernetes resource watched by argo-cd is updated (whenever `resourceVersion` changes), argo-cd will receive an update event and trigger a refresh of the owner applications, if any. Resources that are updated frequently by controllers, such as ExternalSecret, will always have a new `resourceVersion` when information in the `status` field, such as `lastRefreshTime`, `lastEvaluation`, `reconciledAt` are updated. Although the resources are different, these fields should not cause a refresh in argo-cd. A refresh will cause a reconcile operation and even if the diffing evaluates to no changes, it consumes a lot of CPU.

**More details**

The cache from the [argoproj/gitops-engine](https://github.com/argoproj/gitops-engine) only store the `resourceVersion` and does not store the actual resource (except for crds). When a watch event is received, the cache will call OnResourceUpdated handler with the old and new object whenever the `resourceVersion` is updated, but the handler will not be able to do any relevant comparison because the old [Resource](https://github.com/argoproj/gitops-engine/blob/ed70eac8b7bd6b2f276502398fdbccccab5d189a/pkg/cache/resource.go#L14) object does not contain the previous version of the object.

When the argocd OnResourceUpdated handler [receives an event](https://github.com/argoproj/argo-cd/blob/4484ce855834d606b0778a252f69e82034e05592/controller/cache/cache.go#L463), it will simply [trigger/queue](https://github.com/argoproj/argo-cd/blob/4484ce855834d606b0778a252f69e82034e05592/controller/appcontroller.go#L366) a refresh for the application, eventually causing a [reconcile](https://github.com/argoproj/argo-cd/blob/4484ce855834d606b0778a252f69e82034e05592/controller/appcontroller.go#L1529).

This causes a lot of reconciliation for the same application and high CPU usage on both the controller and the repo-server.

**To Reproduce**

1. Update a resource deployed with argoCD by changing a value every 10 seconds (do not change a value that will cause an OutOfSync).
2. Observe the reconcile operation being performed every 10 seconds.

**Expected behavior**

ArgoCD cache should be configurable to ignore certain fields to be evaluated when there is a watch event received. Even if the resourceVersion has changed, the cache should compare the actual objects and if no relevant changes are detected, it should update the cache with the new version **without** triggering the OnResourceUpdated handler.

To save memory, a hash of the manifest with excluded fields could be stored in the `Resource` object.
To configure the cache, 
- argocd could have an additional list of [ResourceIgnoreDifferences](https://github.com/argoproj/argo-cd/blob/4484ce855834d606b0778a252f69e82034e05592/pkg/apis/application/v1alpha1/types.go#LL88C6-L88C31) to ignore fileds before computing the hash.
- Have Lua scripts similar to the [health customization](https://github.com/argoproj/argo-cd/tree/master/resource_customizations) that would receive the object manifest and return the manifest to hash.

**Other considered solution/mitigation**

- Use the new `argocd.argoproj.io/skip-reconcile: "true"` annotations
  - This will block all the reconcile for an application and does not fix the issue.
- Add a new annotation `argocd.argoproj.io/ignore-resource-update: "true"` on each resource that we do not want to trigger a reconcile for the application they belong to
  - This will require the users to add that annotations on multiple objects, probably with MutationWebhook, and adds a lot of burden on the ArgoCD users. It also ignores a resource completely instead of ignoring only certain fields
- Expose configuration for the list of [ignoredRefreshResources](https://github.com/argoproj/argo-cd/blob/4484ce855834d606b0778a252f69e82034e05592/controller/cache/cache.go#L301) 
  - This will ignore a resource completely instead of ignoring only certain fields.

**Screenshots**

Log analysis of which resource update are causing frequent applications refresh (test instance with no load)
image

Total number of App reconciliations (60m) when `timeout.reconciliation: 3m` and expected value should be 20. (test instance with no load)
image

Total number of reconciliations (15m) when `timeout.reconciliation: 5m` and expected value should be around 3. (instance with real load)
image

ArgoCD many application refresh and reconciliation, but no sync.
image

ArgoCD application refresh per controller
image

ArgoCD constant high cpu usage on controller with the most reconcile load
image

**Version**

```shell
argocd-server: v2.5.2+148d8da
  BuildDate: 2022-11-07T16:42:47Z
  GitCommit: 148d8da7a996f6c9f4d102fdd8e688c2ff3fd8c7
  GitTreeState: clean
  GoVersion: go1.18.8
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v4.5.7 2022-08-02T16:35:54Z
  Helm Version: v3.10.1+g9f88ccb
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.18.0
```

**Logs**

```
Paste any relevant application logs here.
```

Related to #9014

### PR DESCRIPTION
**Pull request by agaudreault**
Closes #13534 https://github.com/argoproj/argo-cd/issues/6108 https://github.com/argoproj/argo-cd/issues/13614 https://github.com/argoproj/argo-cd/issues/8471 https://github.com/argoproj/argo-cd/issues/8100 https://github.com/argoproj/argo-cd/issues/7406 https://github.com/argoproj/argo-cd/issues/9014 https://github.com/argoproj/argo-cd/issues/9819

Changes:
- Adding `ignoreResourceUpdates` global configuration to ignore fields before to hash resources.
- Adding `ignoreDifferencesOnResourceUpdates` config to use ignoreDifferences automatically to `ignoreResourceUpdates`.
- Hashing resources **directly** part of an application.
- Filtering out resource updates when hash is the same
- Changed `Refreshing app %s for change in cluster of object %s of type %s/%s` debug log to info to help get statistics and configure `ignoreResourceUpdates`.
  - For Splunk, `msg="Refreshing app*for change*" | rex field=msg "Refreshing app (?<application>\S+) for change in cluster of object (?<resource>\S+) of type (?<type>\S+)" | stats count by application resource type   | sort -count` can be used.
  - EDIT: Now using structured fields.

image
image

This was a result of adding the following config
```yaml
      resource.customizations.ignoreResourceUpdates.all: |
        jsonPointers:
          - /status
```

During business hours, after optimization.
image
image

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

**Comment by jenting** [comment]
This PR looks fantastic. It looks like could fix the app controller CPU high issues.

**Comment by jaideepr97** [comment]
Hi @agaudreault-jive 
I'm not sure I fully understand what problem `ignoreResourceUpdates` is solving that can't be achieved using the existing `ignoreDifferences`, would you mind elaborating on that?

**Comment by agaudreault-jive** [comment]
@jaideepr97 
If your question is related to what is the difference between both: The ignoreDifferences are used to evaluate during the reconcile/refresh of the application if the app should be synchronized or not. Reconcile is the operation consuming the CPU, not sync. So we need to avoid performing the reconcile. Take a look at https://github.com/argoproj/argo-cd/issues/13534 for more details.

If your question is related to why there are 2 different settings and ignoreDifferences could not be reused to skip the reconcile as well: In our case, ignore difference has more configuration than what is necessary for the reconcile optimization. It is also hard/impossible to know what everyone has configured. Having 2 configurations prevents the possibility of conflicts. However, `ignoreDifferencesOnResourceUpdates` is there to simplify the configuration and prevent duplicates.

File: controller/appcontroller.go
Comment (csantanapr) [inline controller/appcontroller.go:368]: Lets try to get cluster name printed in the logs to know which cluster the app is located

Comment (agaudreault-jive) [reply controller/appcontroller.go:368]: I think the comment is more related to https://github.com/agaudreault-jive/argo-cd/blob/704c880ac47c8ace6b04b6d4123a164befc87687/controller/cache/cache.go#L501 where we are still in a specific cluster context.

I can add the cluster server with 
`log.Debugf("Ignoring change in cluster %s of object %s of type %s/%s", clusterCache.GetClusterInfo().Server, resKey, ref.APIVersion, ref.Kind)`

However, I think it would be better if we would be able to use a `clusterCache.GetLogger()` method that would return a log Entry with the current context set on it. This would require changes to the gitops-engine code.

I am not sure that adding the cluster server will provide much value for this log, but I will add it and set a context with WithFields for both logs so we do not need to parse the message 👍

**Comment by donkeyx** [comment]
Agreed, this pr amazing. We are seeing 100% cpu endlessly, with the application-controller monitoring out 2k pods. This seems to be due metric changes on the HPA's.

**Comment by crenshaw-dev** [comment]
@donkeyx are you in a position to run this branch internally and monitor the effects? I'd be happy to help cherry pick these changes to whatever version you're running now.

**Comment by donkeyx** [comment]
> @donkeyx are you in a position to run this branch internally and monitor the effects? I'd be happy to help cherry pick these changes to whatever version you're running now.

I will build and test now in our cluster, unless you have a docker image ready to go? @crenshaw-dev

**Comment by agaudreault-jive** [comment]
@donkeyx for HPA, we had the same issue and I used the following configs to make sure it works with v1 too. You might still see a couple of reconciles due to the ReplicaSet/Pods/Deployments updates, but no more from the HPA.

```
      resource.customizations.ignoreResourceUpdates.all: |
        jsonPointers:
          - /status
      resource.customizations.ignoreResourceUpdates.autoscaling_HorizontalPodAutoscaler:
        |
        jsonPointers:
          - /metadata/annotations/autoscaling.alpha.kubernetes.io~1behavior
          - /metadata/annotations/autoscaling.alpha.kubernetes.io~1conditions
          - /metadata/annotations/autoscaling.alpha.kubernetes.io~1metrics
          - /metadata/annotations/autoscaling.alpha.kubernetes.io~1current-metrics
```

**Comment by agaudreault-jive** [comment]
One "glitch" that I am seeing is when a ReplicaSet is scaled down. When a pod is set to terminating, its health turns to "progressing", the Application health also changes to "progressing". 

However, when the pods "disappear" from the UI, the Application status is not updated and stays "Progressing".

I expect the `OnResourceUpdated` callback to be called when a resource is deleted based on the code with newRes as `nil` and the reconcile to always happens, so I am unsure where the problem can be.

File: controller/appcontroller.go
Comment (crenshaw-dev) [inline controller/appcontroller.go:372]: Maybe add the group, too?

File: controller/cache/cache.go
Comment (crenshaw-dev) [inline controller/cache/cache.go:154]: Why a pointer? Would an empty string suffice for the "does not exist" case?

File: controller/cache/cache.go
Comment (crenshaw-dev) [inline controller/cache/cache.go:184]: I'd add a docstring for this field.

File: controller/cache/cache.go
Comment (crenshaw-dev) [inline controller/cache/cache.go:209]: Wrap with `fmt.Errorf` for easier debugging?

File: controller/cache/cache.go
Comment (crenshaw-dev) [inline controller/cache/cache.go:336]: Func might could use a doc string. Looks like it boils down to "if there's an app name from resource tracking, or if this is itself an app, we should generate a hash."

File: controller/cache/cache.go
Comment (crenshaw-dev) [inline controller/cache/cache.go:504]: Group too?

File: controller/cache/cache.go
Comment (crenshaw-dev) [inline controller/cache/cache.go:505]:
```suggestion
				}).Debug("Ignoring change of object because none of the watched resource fields have changed")
```

File: docs/operator-manual/argocd-cm.yaml
Comment (crenshaw-dev) [inline docs/operator-manual/argocd-cm.yaml:108]: Are `jqPathExpressions` an option here?

File: docs/operator-manual/argocd-cm.yaml
Comment (crenshaw-dev) [inline docs/operator-manual/argocd-cm.yaml:106]: Might also be good to have a resource-specific example, just so folks can be certain that it's an option.

File: util/argo/normalizers/diff_normalizer.go
Comment (crenshaw-dev) [inline util/argo/normalizers/diff_normalizer.go:188]: It's a bit unfortunate we have to detect these this way.

Would you mind adding unit tests for this to avoid regressions if the library changes its strings?

File: util/settings/settings.go
Comment (crenshaw-dev) [inline util/settings/settings.go:831]: Should we document that these are hard-coded optimizations?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Did a quick pass, overall no major concerns. I'm really really excited about this PR!

File: controller/appcontroller.go
Comment (agaudreault-jive) [reply controller/appcontroller.go:372]: The `APIVersion` contains the group already and it seems like the only field available on an ObjectReference object.

File: controller/cache/cache.go
Comment (agaudreault-jive) [reply controller/cache/cache.go:504]: The `APIVersion` contains the group already and it seems like the only field available on an ObjectReference object.

File: util/settings/settings.go
Comment (agaudreault-jive) [reply util/settings/settings.go:831]: already documented in `reconcile.md`. 

I will remove ReconciledAt optimization because it needs other status fields of the Application resource to be configured to work properly. I would prefer to let the users configure this and perhaps have it in the default config in future releases instead of being hardcoded.

File: controller/cache/cache.go
Comment (agaudreault-jive) [reply controller/cache/cache.go:154]: No specific reasons. I will switch to a string.

File: controller/cache/cache.go
Comment (agaudreault-jive) [reply controller/cache/cache.go:340]: I added a doc string, but also kept the detailed comment inside the function because I think it gives valuable information on the implementation and potential improvements/deterioration.

File: docs/operator-manual/reconcile.md
Comment (crenshaw-dev) [inline docs/operator-manual/reconcile.md:50]:
```suggestion
By default, the metadata fields `generation`, `resourceVersion` and `managedFields` are always ignored for all resources.

## Finding Resources to Ignore

The application controller logs when a resource change triggers a refresh. You can use these logs to find
high-churn resource kinds and then inspect those resources to find which fields to ignore.

To find these logs, search for `"Requesting app refresh caused by object update"`. The logs include structured
fields for `api-version` and `kind`.  Counting the number of refreshes triggered, by api-version/kind should
reveal the high-churn resource kinds.

Note that these logs are at the `debug` level. Configure the application-controller's log level to `debug`.
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: controller/cache/info.go
Comment (agaudreault-jive) [reply controller/cache/info.go]: I think the string type is a more universal and clearer type for a hash and it is independent from the algorithm. No strong opinion on that however.

File: util/settings/settings.go
Comment (crenshaw-dev) [inline util/settings/settings.go:786]:
```suggestion
		return nil, fmt.Errorf("failed to get compare options: %w", err)
```

File: util/settings/settings.go
Comment (crenshaw-dev) [inline util/settings/settings.go:791]:
```suggestion
		return nil, fmt.Errorf("failed to get resource overrides: %w", err)
```

File: util/settings/settings.go
Comment (crenshaw-dev) [inline util/settings/settings.go:813]: I hesitate to add this by default, since it impacts diffing. If someone changes this field, I think we want that change reflected in the diff immediately.

```suggestion
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Only one remaining substantial thought.

File: util/settings/settings.go
Comment (agaudreault-jive) [reply util/settings/settings.go:813]: Isn't this field just "metadata"? Usually, people will use this field to evaluate if a specific manager changed another value the object (often something in /spec). 

So i think that even if we exclude the managed fields, another value has to be updated and unless that one is excluded from the user config too, it would still trigger the reconcile.

**Review by crenshaw-dev — APPROVED** [review]

**Comment by csantanapr** [comment]
@agaudreault-jive thanks to and your company for this significant contribution to improving performance 
@crenshaw-dev, as always, the exemplar maintainer taking time to review and discuss the PR 
🎉

**Comment by everythings-gonna-be-alright** [comment]
I Added to my config an additional section:
```  
resource.customizations.ignoreResourceUpdates.batch_CronJob: |
    jsonPointers:
      - /
resource.customizations.ignoreResourceUpdates.kubernetes-client.io_ExternalSecret: |
    jsonPointers:
      - /
```
Then I restarted `argo-cd-argocd-application-controller` but still see in the logs record about ExternalSecrerts and CroJobs:
```
time="2023-07-01T13:16:03Z" level=debug msg="Requesting app refresh caused by object update" api-version=kubernetes-client.io/v1 application=argo-cd/2418 fields.level=1 kind=ExternalSecret name=backup namespace=2418
time="2023-07-01T13:20:16Z" level=debug msg="Requesting app refresh caused by object update" api-version=batch/v1 application=argo-cd/8517 fields.level=1 kind=CronJob name=dynamic-cron-finish-stream namespace=8517
```

**Comment by everythings-gonna-be-alright** [comment]
To enable this feature don't forget to add to your config the next flag:
```
resource.ignoreResourceUpdatesEnabled: "true"
```

**Comment by everythings-gonna-be-alright** [comment]
Actually, after setting `resource.ignoreResourceUpdatesEnabled: "true"`, the CPU usage returns to normal. However, ArgoCD doesn't work properly. All syncs get stuck at this phase:
```
time="2023-07-01T18:17:09Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: 2331)" application=argo-cd/2331
```

On the other hand, when `resource.ignoreResourceUpdatesEnabled: "false"`, adding `resource.customizations.ignoreResourceUpdates.all` doesn't change anything, and the CPU usage remains very high as before.

**Comment by agaudreault-jive** [comment]
@everythings-gonna-be-alright https://github.com/argoproj/argo-cd/pull/14304

until this is merged and cherry-picked, you can used "debug" logs while `resource.ignoreResourceUpdatesEnabled: "false"`, but when you set `resource.ignoreResourceUpdatesEnabled: "true"`, make sure to not have "debug logs.

**Comment by ebuildy** [comment]
Just to be sure, argocd is taking `/status` field in account ? And we must configure to avoid it.

**Comment by agaudreault-jive** [comment]
@ebuildy correct. You have a few example in https://argo-cd.readthedocs.io/en/stable/operator-manual/reconcile

**Pull request merged**
